Add clarification about async vs. sync instruments (#1733)
open-telemetry/opentelemetry-specification: specification/metrics/api.md

* It MUST support at least 1023 characters. Individual language clients can
  decide if they want to support more.

Instruments can be categorized based on whether they are synchronous or
asynchronous:

> <a name="synchronous-instrument"></a>

Reviewer: Why html tags like this? Should use markdown headers if needed.

Reviewer 2: Ping @reyang

Author: It's the same as other tags in this doc (and other spec doc such as the tracing one):
1. we don't want them to be in the top level ToC (since they interrupt the ToC flow)
2. we want to have an anchor

IIRC @bogdandrutu you've asked the same question before?
e.g. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-naming-rule

https://github.com/open-telemetry/opentelemetry-specification/pull/1578#discussion_r602575017

Reviewer 2: @bogdandrutu Please fill an issue to follow up this if this is still a problem, etc. Thanks!

* It MUST support at least 1023 characters. Individual language clients can
  decide if they want to support more.

Instruments can be categorized based on whether they are synchronous or
asynchronous:

> * Synchronous instruments (e.g. [Counter](#counter)) are meant to be invoked by

Reviewer 3: ```suggestion
* Synchronous instruments (e.g. [Counter](#counter)) are meant to be invoked directly by
```

Author: Taken. Thanks!

  the user (e.g. a library developer or an application developer). For example,
  an HTTP client could use a Counter to record the number of bytes it has
  received. [Measurements](#measurement) recorded by synchronous instruments can
  be associated with [Baggage](../baggage/api.md) and
  [Context](../context/context.md).
* Asynchronous instruments (e.g. [Asynchronous Gauge](#asynchronous-Gauge)) give
  the user a way to register callback function, and the callback function will

>   only be invoked upon collection. For example, an embedded software could use

Reviewer 3: ```suggestion
  only be invoked upon collection. For example, a piece of embedded software could use
```

Author: Updated.

  an asynhronous gauge to collect the temperature from a sensor every 15
  seconds, which means the callback function will only be invoked every 15
  seconds. [Measurements](#measurement) recorded by asynchronous instruments
  cannot be associated with [Baggage](../baggage/api.md) and
  [Context](../context/context.md).

Please note that the term _synchronous_ and _asynchronous_ have nothing to do

> with [asynchronous

Reviewer 3: ```suggestion
with the [asynchronous
```

Author: Updated, thanks!

* It MUST support at least 1023 characters. Individual language clients can
  decide if they want to support more.

Instruments can be categorized based on whether they are synchronous or
asynchronous:

<a name="synchronous-instrument"></a>

> * Synchronous instruments (e.g. [Counter](#counter)) are meant to be invoked

Reviewer 4: Nit:  `meant to be invoked inline with application/business processing logic, within the scope of [Context](../context/context.md).`

Baggage  (as specififed) is available via a context. I think you can call it out below as an additional thing, but the key here is;
- The logic is inline with the mainline processing of the application (not sheddable when under load)
- The logic can have automagical attachment of context if desired (and all the fun contexty things).

Author: Please take another look, I believe this has been addressed.

  inline with application/business processing logic. For example, an HTTP client
  could use a Counter to record the number of bytes it has received.
  [Measurements](#measurement) recorded by synchronous instruments can be
  associated with [Baggage](../baggage/api.md) and

>   [Context](../context/context.md).

Reviewer 3: this reads a little funny. Baggage is in the context, as is the current span. If you're going to call out Baggage, I would also call out the Span.

Author: This is a good callout. I agree and I've removed Baggage here (the API spec).

In the SDK spec we will need to call out the difference between Baggage and Context (e.g. the SDK user might specify that he/she wants dimension "a" to come from the Baggage, and dimension "b" to come from the Context).

Reviewer 3: I don't think we'd want anything to come directly from the Context via the SDK. Unless it's a span attribute, of course. The keys to objects in the Context should be private, so the Context couldn't access them.

Author: > The keys to objects in the Context should be private, so the SDK couldn't access them.

+1

The longer version would be: sync instruments can be associated with `Context`, which means the user could configure additional dimensions that come from the Baggage, Span (including span attributes, name/kind/etc.), and anything that we added later to the Context that can be publicly accessed.

I think the key here is to clarify that async instruments are "contextless", so I picked the word "can be associated" instead "have access to". Sounds reasonable?

Reviewer 3: Perfect, thanks!

<a name="asynchronous-instrument"></a>

* Asynchronous instruments (e.g. [Asynchronous Gauge](#asynchronous-gauge)) give
  the user a way to register callback function, and the callback function will
  only be invoked upon collection. For example, a piece of embedded software

>   could use an asynhronous gauge to collect the temperature from a sensor every

Reviewer 3: ```suggestion
  could use an asynchronous gauge to collect the temperature from a sensor every
```
